llvm/lib/IR/DebugProgramInstruction.cpp

```diff
@@ -41,6 +41,36 @@ DPValue::DPValue(Metadata *Location, DILocalVariable *DV, DIExpression *Expr,
 
 void DPValue::deleteInstr() { delete this; }
 
+DPValue *DPValue::createDPValue(Value *Location, DILocalVariable *DV,
+                                DIExpression *Expr, const DILocation *DI,
+                                Instruction *InsertBefore) {
+  auto *NewDPValue = new DPValue(ValueAsMetadata::get(Location), DV, Expr, DI,
+                                 LocationType::Value);
+  if (InsertBefore)
+    InsertBefore->getParent()->insertDPValueBefore(NewDPValue,
+                                                   InsertBefore->getIterator());
```

Comment:
Do you think it's worth having an iterator-taking version too, to match the equivalent insertBefore with BasicBlock::iterator? We don't do before-or-after debuginfo shenanigans here, but that might be necessary if you're looking to have source-equivalence between dbg.value and DPValue mode implementations of code.

Reply:
Perhaps - there might be performance concerns here though: allowing `BasicBlock::iterator` to be passed as an argument here would require `DebugProgramInstruction.h` to `#include "llvm/IR/Instruction.h"`, which _might_ impact performance (including `DebugInfoMetadata.h` here caused a `+2.10%` compile time increase entirely by itself!). I agree that we should probably take a BasicBlock::iterator if possible though; would you consider it acceptable to simply remove the `Instruction *InsertBefore` version (the `DPValue* InsertBefore` is the one that is actually _needed_ for these patches, this overload is optional) and add an iterator-taking version when/if it's confirmed to work, or some other workaround is established?

Reply:
You miiiggghht be able to predeclare it as `SymbolTableList<Instruction, ilist_iterator_bits<true>>::iterator;` without needing a body definition for Instruction.

Removing and adding when necessary seems alright -- hopefully we can avoid it being necessary.

Reply:
I tried that first, but Instruction.h defines a template specialization for Instruction, so unfortunately the predeclaration doesn't work.